### PR TITLE
Cond functionalized

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90632

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Related https://github.com/pytorch/pytorch/issues/50688 ?

File: aten/src/ATen/functorch/ADInterpreters.cpp [hunk 1]
```diff
@@ -115,7 +116,7 @@ static void autogradBasedTransformSendToNext(
     // if (c10::show_dispatch_trace_enabled()) {
     //   std::cout << "wrap " << current_level << std::endl;
     // }
-    return makeTensorWrapper(tensor, current_level, is_immutable);
+    return makeTensorWrapper(tensor, current_level, is_immutable, life_handle);
```
Comment (Skylion007) [inline aten/src/ATen/functorch/ADInterpreters.cpp:119]: All these lower end wrappers take a const ref, for std::shared_ptr, I would recommend just passing by value and judiciously calling when passing it to the next call of makeTensorWrapper.

Comment (Skylion007) [reply aten/src/ATen/functorch/ADInterpreters.cpp:119]:
```suggestion
    return makeTensorWrapper(tensor, current_level, is_immutable, std::move(life_handle));
```

File: aten/src/ATen/functorch/TensorWrapper.cpp [hunk 2]
```diff
@@ -58,19 +58,28 @@ void dumpTensorCout(const Tensor& tensor) {
...
   auto keys_to_propagate = kKeysToPropagateToWrapper | DispatchKeySet({
       DispatchKey::AutogradCPU, DispatchKey::AutogradCUDA, DispatchKey::AutogradXLA});
   auto key_set = getKeysToPropagateToWrapper(tensor, keys_to_propagate);
   key_set = key_set.add(DispatchKey::FuncTorchGradWrapper);
-  if (should_be_alive) {
-    return c10::make_intrusive<TensorWrapper>(key_set, tensor, level, getLifeHandleForLevel(level));
-  } else {
-    return c10::make_intrusive<TensorWrapper>(key_set, tensor, level, std::make_shared<bool>(false));
-  }
+  return c10::make_intrusive<TensorWrapper>(key_set, tensor, level, life_handle);
```
Comment (Skylion007) [inline aten/src/ATen/functorch/TensorWrapper.cpp:66]:
```suggestion
c10::intrusive_ptr<TensorWrapper> makeTensorWrapperPtr(const Tensor& tensor, int64_t level, std::shared_ptr<bool> life_handle) {
  auto keys_to_propagate = kKeysToPropagateToWrapper | DispatchKeySet({
      DispatchKey::AutogradCPU, DispatchKey::AutogradCUDA, DispatchKey::AutogradXLA});
  auto key_set = getKeysToPropagateToWrapper(tensor, keys_to_propagate);
  key_set = key_set.add(DispatchKey::FuncTorchGradWrapper);
  return c10::make_intrusive<TensorWrapper>(std::move(key_set), tensor, level, std::move(life_handle));
```

File: aten/src/ATen/functorch/TensorWrapper.cpp [hunk 3]
```diff
@@ -58,19 +58,28 @@ void dumpTensorCout(const Tensor& tensor) {
...
+      level,
+      is_immutable,
+      getLifeHandleForLevel(level));
+}
+
+Tensor makeTensorWrapper(
+    const Tensor& tensor,
+    int64_t level,
+    bool is_immutable,
+    const std::shared_ptr<bool>& life_handle) {
```
Comment (Skylion007) [inline aten/src/ATen/functorch/TensorWrapper.cpp:82]: Change this to by value wiht std::move as well.

**Pull request closed without merging**